### PR TITLE
Binomial with vectorized total count

### PR DESCRIPTION
**Pull request by neerajprad**
Based on a suggestion from @fritzo prompted by https://github.com/uber/pyro/issues/675, this attempts to restrict same `total_count` only to the `sample` and `enumerate_support` methods for the Binomial distribution, so that we can score samples with vectorized `total_count`. Happy to hear suggestions if there is a non-hacky way to implement sampling for vectorized `total_count` (numpy has this!).

### CODE REVIEW

File: torch/distributions/binomial.py [hunk 1]
```diff
@@ -80,15 +77,22 @@ def probs(self):
...
 
+    def _get_homogeneous_count(self):
+        total_count = int(self.total_count.max().item())
+        if not self.total_count.min() == total_count:
+            raise NotImplementedError("Inhomogeneous total count not supported by method.")
+        return total_count
+
     def sample(self, sample_shape=torch.Size()):
-        shape = self._extended_shape(sample_shape) + (self.total_count,)
+        total_count = self._get_homogeneous_count()
```
Comment (fritzo) [inline torch/distributions/binomial.py]: Gosh I guess since we're already summing `Bernoulli`s we could draw inhomogeneous samples via
```py
def sample(self, sample_shape=torch.Size()):
    with torch.no_grad():
        max_count = self.total_count.max().item()
        shape = self._extended_shape(sample_shape) + (max_count,)
        bernoullis = torch.bernoulli(self.probs.unsqueeze(-1).expand(shape))
        if self.total_count.min() != max_count:
            arange = torch.arange(max_count, out=self.total_count.new_empty(max_count))
            bernoullis *= (arange < self.total_count.unsqueeze(-1)).type_as(bernoullis)
        return bernoullis.sum(dim=-1)
```
WDYT?

Comment (neerajprad) [reply torch/distributions/binomial.py]: I was thinking about it. :) It may lead to us to creating some large intermediate tensors, but it should work fine! Let me make that change and then we can get in @apaszke's opinion.

File: torch/distributions/binomial.py [hunk 2]
```diff
@@ -81,14 +82,19 @@ def param_shape(self):
...
     def sample(self, sample_shape=torch.Size()):
-        shape = self._extended_shape(sample_shape) + (self.total_count,)
+        max_count = int(self.total_count.max().item())
+        shape = self._extended_shape(sample_shape) + (max_count,)
         with torch.no_grad():
-            return torch.bernoulli(self.probs.unsqueeze(-1).expand(shape)).sum(dim=-1)
+            bernoullis = torch.bernoulli(self.probs.unsqueeze(-1).expand(shape))
+            if self.total_count.min() != max_count:
+                arange = torch.arange(max_count, out=self.total_count.new_empty(max_count))
+                bernoullis *= (arange < self.total_count.unsqueeze(-1)).type_as(bernoullis)
```
Comment (neerajprad) [inline torch/distributions/binomial.py]: Super neat trick that you came up with, @fritzo. :)

**Comment by neerajprad** [comment]
@fritzo - could you take another look. I added a few tests, and used your suggestion for vectorized sampling over N.

File: torch/distributions/binomial.py [hunk 3]
```diff
@@ -81,14 +80,19 @@ def param_shape(self):
         return self._param.size()
 
     def sample(self, sample_shape=torch.Size()):
-        shape = self._extended_shape(sample_shape) + (self.total_count,)
+        max_count = int(self.total_count.max().item())
```
Comment (fritzo) [inline torch/distributions/binomial.py]: I'd move this into the `no_grad()` context just to be safe.

File: test/test_distributions.py [hunk 4]
```diff
@@ -2239,6 +2251,15 @@ def test_binomial_shape(self):
         self.assertEqual(dist.log_prob(self.tensor_sample_1).size(), torch.Size((3, 2)))
         self.assertRaises(ValueError, dist.log_prob, self.tensor_sample_2)
 
+    def test_binomial_shape_vectorized_n(self):
```
Comment (fritzo) [inline test/test_distributions.py]: It would also be nice to draw a ton of samples and assert that `(sample <= total_count).all()`, just to make sure we got the masking correct. Maybe also numerically test the mean of `Binomial(total_count=torch.tensor([0,1,2,5,10]), probs=0.5)`?

File: test/test_distributions.py [hunk 5]
```diff
@@ -806,6 +818,17 @@ def test_binomial_extreme_vals(self):
         self.assertAlmostEqual(bin1.log_prob(torch.tensor([float(total_count)]))[0], 0, places=3)
         self.assertEqual(float(bin1.log_prob(torch.tensor([float(total_count - 1)])).exp()), 0, allow_inf=True)
 
+    def test_binomial_vectorized_count(self):
+        set_rng_seed(0)
+        total_count = torch.tensor([[4., 7.], [3., 8.]])
+        bin0 = Binomial(total_count, torch.tensor(1.))
```
Comment (fritzo) [inline test/test_distributions.py]: With a `probs` this high, all you need is a single sample 😉

Comment (neerajprad) [reply test/test_distributions.py]: Yup, this is doing an exact match! I am drawing way more samples for `bin1` with `p=0.5` to make sure that the invariance holds.

**Comment by neerajprad** [comment]
cc. @1Reinier

**Comment by neerajprad** [comment]
Sent upstream to https://github.com/pytorch/pytorch/pull/6720.
